File: 2022-python/day14/day14.py
import os

with open(os.path.join(os.path.dirname(__file__), "input")) as fd:
    puzzle_input = fd.read()

SOURCE = (500, 0)
blocked = set()

for line in puzzle_input.splitlines():
    rs = [list(map(int, p.split(","))) for p in line.split(" -> ")]
    for (x1, y1), (x2, y2) in zip(rs, rs[1:]):
        blocked |= set(
            (x, y)
            for x in range(min(x1, x2), max(x1, x2) + 1)
            for y in range(min(y1, y2), max(y1, y2) + 1)
        )

LP = max(el[1] for el in blocked)
part1 = part2 = 0
sp = SOURCE
while True:
    if part1 == 0 and sp[1] >= LP:
        part1 = part2
    x, y = new_sp = sp
    for nx, ny in ((x, y + 1), (x - 1, y + 1), (x + 1, y + 1)):
        if (nx, ny) not in blocked and ny < LP + 2:
            new_sp = (nx, ny)
            break
    if sp == new_sp:
        blocked.add(sp)
        part2 += 1
        if new_sp == SOURCE:
            break
        sp = SOURCE
        continue
    sp = new_sp

print(part1)
print(part2)

File: 2021-scala/src/main/scala/Day14.scala
import scala.io.Source

object Day14:

  final private val INPUT_DIR = "input/day14/"

  def parseInput(input: String): (String, Map[String, List[String]]) =
    val lines = Source
      .fromFile(INPUT_DIR + input)
      .getLines()
      .toList
    val start = lines(0)
      .strip()
    val map = lines
      .splitAt(2)
      ._2
      .map(l =>
        val sp = l.split(" -> ")
        (sp(0), List(s"${sp(0)(0)}${sp(1)}", s"${sp(1)}${sp(0)(1)}"))
      )
      .toMap
    (start, map)

  def process(
      initialState: List[(String, BigInt)],
      map: Map[String, List[String]],
      ends: String,
      cycles: Int,
  ): Map[Char, BigInt] =
    Range(0, cycles)
      .foldLeft(initialState)((s, i) =>
        s.flatMap(pn => map(pn._1).map((_, pn._2)))
          .groupBy(_._1)
          .mapValues(_.foldLeft(BigInt(0))(_ + _._2))
          .toList
      )
      .flatMap(pn => List((pn._1(0), pn._2), (pn._1(1), pn._2)))
      .groupBy(_._1)
      .mapValues(_.foldLeft(BigInt(0))(_ + _._2))
      .map(cn => if ends.contains(cn._1) then (cn._1, cn._2 + 1) else cn)
      .map(cn => (cn._1, cn._2 / 2))
      .toMap

  def getState(start: String): List[(String, BigInt)] =
    start
      .sliding(2)
      .toList
      .groupBy(identity)
      .mapValues(v => BigInt(v.length))
      .toList

  def getResult(input: String, cycles: Int): BigInt =
    val (start, map) = parseInput(input)
    val state = getState(start)
    val endState =
      process(state, map, s"${start(0)}${start(start.length - 1)}", cycles)
    endState.values.max - endState.values.min

  def part1(input: String): BigInt =
    getResult(input, 10)

  def part2(input: String): BigInt =
    getResult(input, 40)
